### PR TITLE
Disable EMS to fix linked sizing

### PR DESCRIPTION
**Pull request by simularis**
# Pull Request (PR) Description

#29 seems to break the linked sizing feature of modelkit in commercial prototypes. This pull request is meant to restore the ability to do linked-sizing.

- Creates a new parameter `disable_ems`
- Also addresses #59 

#### The problem

To reproduce the error when utilizing linked-sizing:

- User has a cases CSV file with the column `sizing_case` and at least one row refers to another.
- User composes linked-size models with command line `modelkit rake hardsize` or goes straight ahead to simulation with `modelkit rake`. The resulting models look like `instance-hardsize.idf`.
- After running the models, the user may see an error message in `instance-hardsize-out.err` that there is no System:Sizing object, although there are Autosize input fields present.

#### Proposed solution

The proposed solution is to disable the block of EMS logic in `unitary.pxt` added recently, by creating a parameter to guard that block of code. The logic will be enabled by default (`disable_ems` = false), but users can include `disable_ems` = true as an input parameter to turn it off.

### PR Author
- [X] Make sure the PR branch is up to date with main branch at the time of the PR submission
- [X] Craft a succinct title that effectively encapsulates the essence of the pull request, providing a general overview of the proposed changes.
- [ ] Label the PR with at least one of the following: New Measure, Bug, or Feature.
- [X] Provide a concise description of the measure, bug, or feature. Submit one PR per measure.
- [ ] For a new measure, attach a workbook named DEER_EnergyPlus_Modelkit_Measure_list_working.xlsx, containing only rows used for post-processing the measure.
- [x] Add comments in the code when necessary to facilitate the review process.
- [x] Add a comment before the added code, including the author's full name, company, and specifying if it's a bug fix, new measure, or feature.
- [ ] For a new feature or bug, demonstrate the impact on energy consumption for selected cases with justification using plots and descriptions.
- [ ] For a new measure, add a summary table showing total energy consumption per simulated case.
### PR Reviewer
- [ ] Conduct a thorough code review.
- [ ] If the branch is behind the main, merge the branch locally to check for potential conflicts.
- [ ] If a bug, locally reproduce it and compare energy consumptions before and after.
- [ ] Explore creative ways to stress-test the code.
- [ ] Locally check the error file and other outputs.

### CODE REVIEW

File: templates/energyplus/templates/system/unitary.pxt [hunk 1]
```diff
@@ -204,6 +204,8 @@ parameter "sat_temp_max", :default => Autosize
 
 parameter "clear_water_storage_tank", :default => ""  # Name of tank for collecting cooling coil condensate
 
+# Solaris Technical 2024-05-07 for linked sizing
+parameter "disable_ems", :default => false, :type => "Enumeration", :valuelist => "true,false"
```
Comment (simularis) [inline templates/energyplus/templates/system/unitary.pxt:209]: Here we create a new parameter, which will pass through to root prototypes.

File: templates/energyplus/templates/system/unitary.pxt [hunk 2]
```diff
@@ -2852,6 +2854,12 @@ cool_flow_min_frac = 0.67
 end
 %>
 
+!- Value of disable_ems = <%= disable_ems %>
+<%
+# Solaris Technical 2024-05-07
+# Because this airflow EMS logic interferes with Modelkit's linked-sizing feature, create a parameter to disable it.
+if (!disable_ems)
+%>
```
Comment (simularis) [inline templates/energyplus/templates/system/unitary.pxt:2862]: Here we guard the EMS code with an IF-clause.

File: templates/energyplus/templates/zonehvac/hvac-zone.pxt [hunk 3]
```diff
@@ -391,6 +391,9 @@ rule "hvac_type_ideal", :parameters => {:hvac_type => "IDEAL"} do
 
 # disable sub_heat and exh_fan
 
+# Solaris Technical 2024-05-07 for linked sizing
+parameter "disable_ems", :default => false, :type => "Enumeration", :valuelist => "true,false"
+
```
Comment (simularis) [inline templates/energyplus/templates/zonehvac/hvac-zone.pxt:396]: Unitary template is often called from hvac-zone, so need the parameter here, too.

File: templates/energyplus/templates/zonehvac/hvac-zone.pxt [hunk 4]
```diff
@@ -867,7 +870,9 @@ unitary_parameters = [
   :aux_heat_coil_type,
   :aux_heat_coil_cap,
   :aux_heat_coil_eff,
-  :aux_heat_coil_fuel
+  :aux_heat_coil_fuel,
+
+  :disable_ems
```
Comment (simularis) [inline templates/energyplus/templates/zonehvac/hvac-zone.pxt:875]: The unitary_parameters list will feed into the insert unitary statement later.

File: prototypes/ECC/templates/root.pxt [hunk 5]
```diff
@@ -188,6 +188,7 @@ end
 
 group "space_hvac" do
   import "zonehvac/hvac-zone.pxt", :group => "building" do # hardcode building-level defaults
+    include "disable_ems"
     parameter "oa_person", :default => 0['CFM']
```
Comment (simularis) [inline prototypes/ECC/templates/root.pxt:192]: No need to redefine it, get the default value from hvac-zone.

File: prototypes/ECC/templates/root.pxt [hunk 6]
```diff
@@ -713,7 +714,7 @@ for zone in all_zones
     :sys_name => "#{this_hvac_type}-#{sys_name}",
     :clear_water_storage_tank => condensate_collection,
     :clear_water_supply_tank => clear_water_supply_tank,
-    **( this_hvac | this_hvac_zone | space_hvac[zone_type] | space_hvac[zone_sch_group])
+    **( space_hvac.building | this_hvac | this_hvac_zone | space_hvac[zone_type] | space_hvac[zone_sch_group])
```
Comment (simularis) [inline prototypes/ECC/templates/root.pxt:718]: Pass the building-level default parameters for hvac-zone into the template here. This also includes `oa_person` (default 0) and `sat_min_dt` (unused by template).

File: prototypes/ECC/templates/root.pxt [hunk 7]
```diff
@@ -814,7 +815,7 @@ for zone in irregular_zones
     :zone_name => zone_name,
     :zone_area => zone_area,
     :sys_name => sys_name,
-    **(zone_hvac_group)
+    **(space_hvac.building | zone_hvac_group)
```
Comment (simularis) [inline prototypes/ECC/templates/root.pxt:818]: Some prototypes have irregular_zones. Pass the building-level parameters for hvac-zone here, also.

**Review by amine-lazrak — CHANGES_REQUESTED** [review]
Thank you @simularis for adding the new parameter. Could you please merge the updated main branch with this one for an easier integration after review. I only had a minor comment that you will find below.

**Comment by simularis** [comment]
@amine-lazrak, thank you for the review and catching the hvac_type_ideal rule. I made the requested changes.

**Pull request merged**